### ISSUE TITLE
Calendar 4.7.9 prevents sharing files anonymously

### ISSUE THREAD

**Issue by rmbleeker** [nextcloud/calendar#6120]
### Steps to reproduce

1. Share a note using a public link
2. Upgrade calendar app to version 4.7.9
3. Get an Internal Server Error

### Expected behavior

Expected the public link to work as it used to.

### Actual behaviour

Internal Server Error

### Calendar app version

4.7.9

### Browser

Firefox 127

### Client operating system

Pop!_OS 22.04 LTS

### Server operating system

Debian 12.6

### Web server

Nginx

### Database engine version

MySQL

### PHP engine version

PHP 8.2

### Nextcloud version

29.0.3

### Updated from an older installed version or fresh install

Updated from an older version

### List of activated apps

```Text
activity: 2.21.1
calendar: 4.7.9
circles: 29.0.0-dev
cloud_federation_api: 1.12.0
comments: 1.19.0
contacts: 6.0.0
contactsinteraction: 1.10.0
cookbook: 0.11.1
dashboard: 7.9.0
dav: 1.30.1
deck: 1.13.1
federatedfilesharing: 1.19.0
federation: 1.19.0
files: 2.1.0
files_downloadlimit: 2.0.0
files_pdfviewer: 2.10.0
files_reminders: 1.2.0
files_sharing: 1.21.0
files_trashbin: 1.19.0
files_versions: 1.22.0
firstrunwizard: 2.18.0
forms: 4.2.4
integration_github: 2.0.7
logreader: 2.14.0
lookup_server_connector: 1.17.0
mail: 3.7.2
nextcloud_announcements: 1.18.0
notes: 4.10.0
notifications: 2.17.0
notify_push: 0.6.12
oauth2: 1.17.0
password_policy: 1.19.0
passwords: 2024.5.20
photos: 2.5.0
previewgenerator: 5.5.0
privacy: 1.13.0
provisioning_api: 1.19.0
recommendations: 2.1.0
related_resources: 1.4.0
richdocuments: 8.4.3
richdocumentscode: 24.4.402
serverinfo: 1.19.0
settings: 1.12.0
sharebymail: 1.19.0
side_menu: 3.12.0
spreed: 19.0.4
support: 1.12.0
survey_client: 1.17.0
systemtags: 1.19.0
tasks: 0.16.0
text: 3.10.1
theming: 2.4.0
twofactor_admin: 4.5.0
twofactor_backupcodes: 1.18.0
twofactor_nextcloud_notification: 3.9.0
twofactor_totp: 11.0.0-dev
twofactor_webauthn: 1.4.0
unroundedcorners: 1.1.3
updatenotification: 1.19.1
user_status: 1.9.0
viewer: 2.3.0
weather_status: 1.9.0
workflowengine: 2.11.0
```

### Nextcloud configuration

```Text
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.stormhooligans.nl"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "enable_previews": true,
        "preview_max_x": 4096,
        "preview_max_y": 4096,
        "appcodechecker": true,
        "log_type": "file",
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "loglevel": 2,
        "logtimezone": "Europe\/Amsterdam",
        "log_rotate_size": 10485760,
        "dbtype": "mysql",
        "version": "29.0.3.4",
        "overwrite.cli.url": "https:\/\/cloud.stormhooligans.nl",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "defaultapp": "files",
        "default_phone_region": "NL",
        "session_lifetime": 86400,
        "htaccess.RewriteBase": "\/",
        "filelocking.enabled": "true",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0
        },
        "installed": true,
        "twofactor_enforced": "false",
        "twofactor_enforced_groups": [],
        "twofactor_enforced_excluded_groups": [
            "admin"
        ],
        "maintenance": false,
        "maintenance_window_start": 3,
        "mysql.utf8mb4": true,
        "theme": "",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "updater.release.channel": "stable",
        "mail_smtpsecure": "ssl",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### Log file

```Text
[index] Error: OCA\Calendar\Service\Appointments\AppointmentConfigService::getAllAppointmentConfigurations(): Argument #1 ($userId) must be of type string, null given, called in /var/www/html/nextcloud/apps/calendar/lib/Service/CalendarInitialStateService.php on line 96 in file '/var/www/html/nextcloud/apps/calendar/lib/Service/Appointments/AppointmentConfigService.php' line 62
	GET /s/DxXBHYdwFpbgdLW
	from xx.xx.xx.xx by -- at 4 jul 2024 10:52:26
```

### Additional info

After downgrading the calendar app to the previous version (4.7.8) the public link is working again.

**Comment by rmbleeker** [comment]
I did not test if this bug also affects regular file sharing with other users, or sharing other files than notes through a public link. I run a self-hosted Nextcloud that I share with family and friends and don't have a testing environment, so my ability to test and debug is limited.

**Comment by mtorromeo** [comment]
I can confirm that it does affect sharing files publicly through links. The code errors out when it's expecting to find a logged in user id.

**Comment by JimmyKater** [comment]
As it seems there could be a problem with the new version of Calendar. Is it possible to remove/disable the actual version from app-store? just to prevent more users running into that bug? 
I think it would be really helpful

**Comment by fredvd** [comment]
This is not only happening when trying to share a note, but also when you want to share an anonymous drop file sharing link.

**Comment by rmbleeker** [comment]
I was being conservative with my report and only wanted to report what I experienced myself, but I think it's safe to say that all anonymous sharing, so without any userID involved, is affected by this bug. The error suggests as much.

**Comment by Bobmorton-TX** [comment]
I can confirm that the current calendar 4.7.9 app creates an error on all anonymous file sharing. NC29. disabling the app and the shares throw no interal server error anymore

**Comment by Bobmorton-TX** [comment]
Just installed 4.7.10 this fixed my issues. thanks!!!

**Comment by rmbleeker** [comment]
Is there a way to force Nextcloud to check for new app updates? I can't find a GUI option for it nor an `occ` subcommand. My instance still sees 4.7.9 as the latest version whenever I try to upgrade.

**Comment by rdx-pg** [comment]
There is `occ update:check` but no update to 4.7.10 yet

**Comment by rdx-pg** [comment]
A quick-fix I did would be to change the file `lib/Service/CalendarInitialStateService.php` line 96 similar to https://github.com/nextcloud/calendar/blob/v4.7.10/lib/Service/CalendarInitialStateService.php#L96

**Comment by RAurelian** [comment]
Is it possible that the same error occurs for other plugins as well?

For example we have this after installing the Deck application:
`<html>
<body>

Error | PHP | Optional parameter $userId declared before required parameter $path is implicitly treated as a required parameter at /var/www/nextcloud/apps/deck/lib/Service/ConfigService.php#234
-- | -- | --



</body>
</html>`

**Comment by rmbleeker** [comment]
> A quick-fix I did would be to change the file `lib/Service/CalendarInitialStateService.php` line 96 similar to https://github.com/nextcloud/calendar/blob/v4.7.10/lib/Service/CalendarInitialStateService.php#L96

I would advise against changing the code directly, Nextcloud's code integrity checker might not approve. Instead I would download a tar.gz file with version 4.7.10 of the Calendar app from https://apps.nextcloud.com/apps/calendar/releases?platform=29 and use that.

- disable the calendar app in Nextcloud
- put Nextcloud in maintenance mode
- optionally stop your webserver and php runtime
- delete the `calendar` directory in `.../nextcloud/apps/`
- extract the tar.gz into the `.../nextcloud/apps/` directory
- start the webserver and php runtime if you stopped them
- disable maintenance mode on Nextcloud
- enable the calendar app in the Nextcloud GUI
- check if calendar is indeed at version 4.7.10 and if anonymous shares still function correctly

**Comment by rmbleeker** [comment]
> Is it possible that the same error occurs for other plugins as well?
> 
> For example we have this after installing the Deck application: `Error | PHP | Optional parameter $userId declared before required parameter $path is implicitly treated as a required parameter at /var/www/nextcloud/apps/deck/lib/Service/ConfigService.php#234`

That's a different error that I've seen in my logs as well, but as far as I can tell that doesn't prevent you from creating anonymous file shares. I know the Deck developers have already fixed this bug in the code, a new version with that code just hasn't been released yet.

**Comment by RAurelian** [comment]
Luckily the official Calendar app was already updated and when updating it in Nextcloud it fixed the issue with anonymous sharing.

**Comment by RAurelian** [comment]
> > Is it possible that the same error occurs for other plugins as well?
> > For example we have this after installing the Deck application: `Error | PHP | Optional parameter $userId declared before required parameter $path is implicitly treated as a required parameter at /var/www/nextcloud/apps/deck/lib/Service/ConfigService.php#234`
> 
> That's a different error that I've seen in my logs as well, but as far as I can tell that doesn't prevent you from creating anonymous file shares. I know the Deck developers have already fixed this bug in the code, a new version with that code just hasn't been released yet.

Thank you - then we`ll wait for the new Deck update

**Comment by miaulalala** [comment]
Is it fixed for everyone in 4.7.10 now?

**Comment by dvdsnchzIT** [comment]
Hello,

I have the problem with nextcloud 28.0.7, calendar 4.7.10 and deck 1.12.3.

It works from my internal network, but not from the outside.
